### PR TITLE
[skip ci]publish npm packages to npmjs

### PR DESCRIPTION
**Pull request by PengTian0**
The PR implements 2 functions:

- Generate the version of the npm packages to be published.
- Publish to npm package to registry of npm
    "latest" tag with official release packages.
    "ci-release" tag with nightly release packages.

**Test**:
http://rackhdci.lss.emc.com/job/Experimental-Jobs/job/Release_NPM

If users want to install the latest stable release, they need to run ```npm install package```
If users want to install the latest nightly release, they need to run ```npm install package@ci-release```

The npm registry: https://www.npmjs.com/~rackhd
You can get the details about a npm package by running ```npm view package```

### CODE REVIEW

**Comment by PengTian0** [comment]
test this please

**Comment by PengTian0** [comment]
test this please

**Comment by PengTian0** [comment]
@panpan0000 @changev @anhou @yyscamper

**Comment by PengTian0** [comment]
test this please

**Review by anhou — APPROVED** [review]

**Comment by PengTian0** [comment]
@panpan0000 @changev  @yyscamper

File: build-release-tools/application/release_npm_packages.py [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+
+"""
+This is a command line program that makes a rackhd release to npmjs.
+
+Usage:
+./on-build-config/build-release-tools/HWIMO-BUILD on-build-config/build-release-tools/application/release_npm_packages.py \
+--build-directory b/ \
+--manifest-file rackhd-devel \
+--git-credential https://github.com,GITHUB \
+--npm-credential registry.npmjs.org,6a2d8de2-dfce-4f59-a866-20f1baa394bd \
```
Comment (panpan0000) [inline build-release-tools/application/release_npm_packages.py:12]: is this credential a real one ? if yes, we don't want to expose it ?

File: build-release-tools/application/release_npm_packages.py [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+        if os.path.isdir(repo_dir):
+            for filename in os.listdir(repo_dir):
+                if filename == "package.json":
+                    packages.append(dirname)
+    return packages
+
+def compute_packages_version(build_dir, is_official_release):
+    '''
+    compute the version of packages under the build_dir
+    return: a dict like: {"pacakge name": "version"}
```
Comment (panpan0000) [inline build-release-tools/application/release_npm_packages.py]: pacakge --> package

File: build-release-tools/application/release_npm_packages.py [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+    '''
+    Find all the packages which contains package.json under the directory specified by the parameter build_dir
+    '''
+    packages = []
+    for dirname in os.listdir(build_dir):
+        repo_dir = os.path.join(build_dir,dirname)
+        if os.path.isdir(repo_dir):
+            for filename in os.listdir(repo_dir):
+                if filename == "package.json":
+                    packages.append(dirname)
```
Comment (panpan0000) [inline build-release-tools/application/release_npm_packages.py:99]: then break ?

File: build-release-tools/application/release_npm_packages.py [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+    npm_packages.remove("di.js")
+    for package in npm_packages:
+        try:
+            package_dir = os.path.join(build_dir, package)
+            npm = NPM(npm_registry, npm_token)
+            print "starting to publish npm package {0}".format(package)
+            if is_official_release:
+                npm.publish_package(package_dir)
+            else:
+                npm.publish_package(package_dir, tag="ci-release")
```
Comment (panpan0000) [inline build-release-tools/application/release_npm_packages.py:198]: what does the "tag" means?
can you show us where the tag is in npm registry ?

**Review by panpan0000 — COMMENTED** [review]
can you add something example you already uploaded in https://www.npmjs.com ? so that people can know what it will like ...
and moreover, I believe you will have a story to update the readthedoc , to tell people how to install

File: build-release-tools/application/release_npm_packages.py (hunk 1 above)
Comment (PengTian0) [reply build-release-tools/application/release_npm_packages.py:12]: It is not a real one.

File: build-release-tools/application/release_npm_packages.py (hunk 3 above)
Comment (PengTian0) [reply build-release-tools/application/release_npm_packages.py:99]: Thanks!
Fixed

File: build-release-tools/application/release_npm_packages.py (hunk 4 above)
Comment (PengTian0) [reply build-release-tools/application/release_npm_packages.py:198]: The tag "ci-release" is used to install the latest version of nightly build.
If users want to install the latest stable release, they need to run ```npm install package```
If users want to install the latest nightly release, they need to run ```npm install package@ci-release```

The npm registry: https://www.npmjs.com/~rackhd
You can get the details about a npm package by running ```npm view package```

File: build-release-tools/application/release_npm_packages.py (hunk 3 above)
Comment (PengTian0) [inline build-release-tools/application/release_npm_packages.py]: Thanks!
Fixed.

File: build-release-tools/application/release_npm_packages.py (hunk 2 above)
Comment (PengTian0) [reply build-release-tools/application/release_npm_packages.py]: Thanks!
Fixed.

File: build-release-tools/application/release_npm_packages.py (hunk 4 above)
Comment (panpan0000) [reply build-release-tools/application/release_npm_packages.py:198]: :+1:

File: build-release-tools/application/release_npm_packages.py (hunk 1 above)
Comment (ChALkeR) [reply build-release-tools/application/release_npm_packages.py:12]: @PengTian0 It is a real one.
It looks like those credentials are for a personal test account, and not the actual `rackhd` one, though.

Comment (ChALkeR) [inline build-release-tools/application/release_npm_packages.py]: It looks like those credentials are for a personal test account, and not the actual `rackhd` one, though.

**Pull request merged**